contrib/colab/DC-GAN.ipynb

```diff
@@ -0,0 +1,557 @@
+        "\n",
+        "The second place, where detach() is used is when a numpy() call is to be made to tensor (for plotting purposes). Pytorch also requires that requires_grad should not be true on these tensor. (Ref: `\n",
+        "RuntimeError: Can't call numpy() on Variable that requires grad. Use var.detach().numpy() instead.`)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "colab_type": "code",
+        "id": "OFzyeN_GQhxk",
+        "colab": {}
+      },
+      "source": [
+        "def train_gan(rank):\n",
+        "    torch.manual_seed(1)\n",
+        "    \n",
+        "    if not xm.is_master_ordinal():\n",
+        "        # Barrier: Wait until master is done downloading\n",
+        "        xm.rendezvous('download_only_once')\n",
+        "    # Dataset\n",
+        "    data = mnist_data()\n",
+        "    if xm.is_master_ordinal():\n",
+        "        # Master is done, other workers can proceed now\n",
+        "        xm.rendezvous('download_only_once')\n",
+        "    train_sampler = torch.utils.data.distributed.DistributedSampler(\n",
+        "        data,\n",
+        "        num_replicas=xm.xrt_world_size(),\n",
+        "        rank=xm.get_ordinal(),\n",
+        "        shuffle=True)\n",
+        "    \n",
+        "\n",
+        "    # Create loader with data, so that we can iterate over it\n",
+        "    #train_loader = torch.utils.data.DataLoader(data, batch_size=100, shuffle=True)\n",
+        "    train_loader = torch.utils.data.DataLoader(\n",
+        "      data,\n",
+        "      batch_size=FLAGS['batch_size'],\n",
+        "      sampler=train_sampler,\n",
+        "      num_workers=FLAGS['num_workers'],\n",
+        "      drop_last=True)\n",
+        "\n",
+        "    # Num batches\n",
+        "    num_batches = len(train_loader)\n",
+        "    \n",
+        "    device = xm.xla_device()\n",
+        "    \n",
+        "    generator = GenerativeNet().to(device)\n",
+        "    generator.apply(init_weights)\n",
+        "\n",
+        "    discriminator = DiscriminativeNet().to(device)\n",
+        "    discriminator.apply(init_weights)\n",
+        "    \n",
+        "    # Optimizers\n",
+        "    d_optimizer = Adam(discriminator.parameters(), lr=FLAGS['disc_learning_rate'], betas=(0.5, 0.999))\n",
+        "    g_optimizer = Adam(generator.parameters(), lr=FLAGS['gen_learning_rate'], betas=(0.5, 0.999))\n",
+        "\n",
+        "    # Number of epochs\n",
+        "    num_epochs = FLAGS['num_epochs'] \n",
+        "    # Loss function\n",
+        "    loss = nn.BCELoss()\n",
+        "    \n",
+        "    num_test_samples = 24\n",
+        "    test_noise = noise(num_test_samples, device)\n",
```

Comment:
can we move this closer to the line where you plot test_noise?

`    xm.do_on_ordinals(plot_results, generator(test_noise).detach().cpu(), (0,))`

Reply:
But that would mean generating test noise over and over again. which is something we dont want. we want to generate a test_noise once and ideally test the generator on what it does with it. We are doing plot image once at the end because I could not find a way to dynamically display the images at the end of each epoch.

Reply:
Oh I think you mean after the end of all the epochs?

Reply:
Moving it.

Reply:
yeah yeah, just before plotting. I had to figure out why you defined this variable, I was confused. If it was immediately used, it would read better.

Reply:
Done.